### PR TITLE
Fix bug where in-place Hamiltonian addition was not updating metadata

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The following was failing,

```python
H = qml.PauliZ(0) + qml.PauliZ(1)
H += qml.PauliZ(2)
```

since the `Hamiltonian.__iadd__()` method was not correctly updating information such as wires.

**Description of the Change:** Adds logic to `iadd` to make this change.

**Benefits:** The above now works correctly.

**Possible Drawbacks:** We could better code up this class to avoid duplicating this logic.

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/ops/qubit/hamiltonian.py [hunk 1]
```diff
@@ -567,7 +567,7 @@ def __mul__(self, a):
         r"""The scalar multiplication operation between a scalar and a Hamiltonian."""
         if isinstance(a, (int, float)):
             self_coeffs = copy(self.coeffs)
-            coeffs = qml.math.multiply(a, self_coeffs)
+            coeffs = qml.math.multiply(qml.math.cast_like([a], self_coeffs), self_coeffs)
```
Comment (antalszava) [inline pennylane/ops/qubit/hamiltonian.py:570]: Hi @josh146, just wanted to double-check: are the the changes made to `__mul__` and `__imul__` intentional?

They were part of a bug fix PR (#2376) where e.g., floating scalars were cast to integers due to the use of `qml.math.cast_like`.

Comment (josh146) [reply pennylane/ops/qubit/hamiltonian.py:570]: Oh! I did not change these locally, I wonder why this happened.

Comment (josh146) [reply pennylane/ops/qubit/hamiltonian.py]: have removed these!

**Comment by antalszava** [comment]
@josh146 oddly enough, not able to reproduce this on `master`:
```python
In [1]: import pennylane as qml
   ...: 
   ...: H = qml.PauliZ(0) + qml.PauliZ(1)
   ...: H += qml.PauliZ(2)

In [3]: H.terms()
Out[3]: ((1, 1, 1), [PauliZ(wires=[0]), PauliZ(wires=[1]), PauliZ(wires=[2])])
```
What was the original error raised?

Also, it seems that some tests started failing.

**Comment by josh146** [comment]
> What was the original error raised?

`H.wires()` will be incorrect, and as a result, `qml.matrix(H)` will fail :)

**Comment by antalszava** [comment]
Hi @josh146, would you like us to have a look at what's going on here?

**Comment by antalszava** [comment]
[sc-18750]

**Comment by josh146** [comment]
@antalszava apologies, I think I have run out of time in the near term to come back to this!

**Comment by Jaybsoni** [comment]
This is weird, I think I fixed this bug last quarter ..? https://github.com/PennyLaneAI/pennylane/pull/2738

File: pennylane/ops/qubit/hamiltonian.py [hunk 2]
```diff
@@ -582,21 +582,26 @@ def __sub__(self, H):
...
             self._ops.append(H)
-            self.simplify()
-            return self
 
-        raise ValueError(f"Cannot add Hamiltonian and {type(H)}")
+        # HOTFIX: repeat some of the __init__ logic
+        self._hyperparameters = {"ops": self._ops}
+        self._wires = qml.wires.Wires.all_wires([op.wires for op in self.ops], sort=True)
+        self._grouping_indices = None
+        self.data = [self._coeffs[i] for i in range(qml.math.shape(self._coeffs)[0])]
```
Comment (Jaybsoni) [inline pennylane/ops/qubit/hamiltonian.py:602]: Ah Yes! We need to update all of the meta data, not just the wires. 
In this case we could get rid of the wires being updated in the `simplify` method as this is now redundant

**Comment by Jaybsoni** [comment]
Upon closer inspection it seems that this has already been solved. Possibly by another PR or as part of a different contribution. Any concerns with closing this issue @josh146, @antalszava ?

**Comment by josh146** [comment]
@Jaybsoni if this PR is no longer valid happy to close it :)

**Pull request closed without merging**